File: java/com/example/backend/service/ProductService.java
package com.example.backend.service;

import com.example.backend.model.Product;
import com.example.backend.repository.ProductRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;
import org.springframework.data.domain.Sort;
import org.springframework.stereotype.Service;

import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Optional;

@Service
public class ProductService {

    @Autowired
    private ProductRepository productRepository;

    // Tüm ürünleri sayfalama ve filtreleme ile getir
    public Map<String, Object> getProducts(int page, int size, String sort, String category, String search) {
        Sort.Direction direction = Sort.Direction.ASC;
        String property = "id";
        
        if (sort != null && !sort.isEmpty()) {
            String[] sortParams = sort.split(",");
            if (sortParams.length > 1) {
                direction = sortParams[1].equalsIgnoreCase("desc") ? Sort.Direction.DESC : Sort.Direction.ASC;
            }
            property = sortParams[0];
        }
        
        Pageable pageable = PageRequest.of(page, size, Sort.by(direction, property));
        Page<Product> productPage;
        
        if (search != null && !search.isEmpty() && category != null && !category.isEmpty()) {
            try {
                Long categoryId = Long.parseLong(category);
                productPage = productRepository.searchProductsByCategoryId(search, categoryId, pageable);
            } catch (NumberFormatException e) {
                productPage = productRepository.findByCategorySlugAndActive(category, pageable);
            }
        } else if (search != null && !search.isEmpty()) {
            productPage = productRepository.searchProducts(search, pageable);
        } else if (category != null && !category.isEmpty()) {
            try {
                // Sayısal ID ise 
                Long categoryId = Long.parseLong(category);
                // Alt kategorileri de içerecek şekilde ürünleri getir
                productPage = productRepository.findByCategoryHierarchy(categoryId, pageable);
            } catch (NumberFormatException e) {
                // Slug ise
                // Alt kategorileri de içerecek şekilde ürünleri getir
                productPage = productRepository.findByCategorySlugHierarchy(category, pageable); 
            }
        } else {
            productPage = productRepository.findAll(pageable);
        }
        
        Map<String, Object> response = new HashMap<>();
        response.put("items", productPage.getContent());
        response.put("total", productPage.getTotalElements());
        
        return response;
    }

    // Öne çıkan ürünleri getir
    public List<Product> getFeaturedProducts() {
        return productRepository.findByActiveAndFeaturedTrue(true);
    }

    // Tek bir ürün detayını getir
    public Product getProduct(Long id) {
        return productRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Product not found with id: " + id));
    }

    // Kategoriye göre ürünleri getir - alt kategorileri içerir
    public List<Product> getProductsByCategory(Long categoryId) {
        try {
            // Önce daha basit sorguyu dene
            List<Product> products = productRepository.findSimpleProductsByCategoryHierarchy(categoryId);
            if (products != null && !products.isEmpty()) {
                return products;
            }
            
            // Eğer başarısız olursa veya sonuç yoksa, daha karmaşık sorguyu kullan
            Pageable pageable = PageRequest.of(0, 100);
            return productRepository.findByCategoryHierarchy(categoryId, pageable).getContent();
        } catch (Exception e) {
            // Hata durumunda sadece kategori ID'si ile ürünleri getir
            System.out.println("Kategori hiyerarşi sorgusu başarısız: " + e.getMessage());
            Pageable pageable = PageRequest.of(0, 100);
            return productRepository.findByCategoryIdAndActive(categoryId, true, pageable).getContent();
        }
    }
    
    // Kategori slug'ına göre ürünleri getir - alt kategorileri içerir
    public List<Product> getProductsByCategorySlug(String slug) {
        System.out.println("DEBUG: Looking for products with category slug: " + slug);
        try {
            // Önce daha basit sorguyu dene
            List<Product> products = productRepository.findSimpleProductsByCategorySlugHierarchy(slug);
            System.out.println("DEBUG: Simple query result count: " + (products != null ? products.size() : 0));
            if (products != null && !products.isEmpty()) {
                return products;
            }
            
            // Eğer başarısız olursa veya sonuç yoksa, daha karmaşık sorguyu kullan
            Pageable pageable = PageRequest.of(0, 100);
            List<Product> pageProducts = productRepository.findByCategorySlugHierarchy(slug, pageable).getContent();
            System.out.println("DEBUG: Complex query result count: " + (pageProducts != null ? pageProducts.size() : 0));
            return pageProducts;
        } catch (Exception e) {
            // Hata durumunda sadece kategori slug'ı ile ürünleri getir
            System.out.println("DEBUG: Category slug hierarchy query failed: " + e.getMessage());
            e.printStackTrace();
            Pageable pageable = PageRequest.of(0, 100);
            List<Product> fallbackProducts = productRepository.findByCategorySlugAndActive(slug, pageable).getContent();
            System.out.println("DEBUG: Fallback query result count: " + (fallbackProducts != null ? fallbackProducts.size() : 0));
            return fallbackProducts;
        }
    }

    // İndirimli ürünleri getir
    public List<Product> getDiscountedProducts() {
        return productRepository.findDiscountedProducts();
    }

    // En çok satanları getir
    public List<Product> getBestSellers() {
        return productRepository.findBestSellers();
    }

    // Yeni ürünleri getir
    public List<Product> getNewArrivals() {
        Pageable pageable = PageRequest.of(0, 10);
        return productRepository.findByActiveOrderByCreatedAtDesc(true, pageable).getContent();
    }

    // Slug ile ürün getir
    public Product getProductBySlug(String slug) {
        return productRepository.findBySlug(slug)
                .orElseThrow(() -> new RuntimeException("Product not found with slug: " + slug));
    }

    // Yeni ürün ekle
    public Product addProduct(Product product) {
        return productRepository.save(product);
    }

    // Ürün güncelle
    public Product updateProduct(Long id, Product productDetails) {
        Product product = getProduct(id);
        
        product.setName(productDetails.getName());
        product.setDescription(productDetails.getDescription());
        product.setPrice(productDetails.getPrice());
        product.setStock(productDetails.getStock());
        product.setImageUrl(productDetails.getImageUrl());
        
        if (productDetails.getCategory() != null) {
            product.setCategory(productDetails.getCategory());
        }
        
        return productRepository.save(product);
    }

    // Ürün sil
    public void deleteProduct(Long id) {
        Product product = getProduct(id);
        productRepository.delete(product);
    }

    public List<Product> getProductsByCategoryHierarchy(Long categoryId) {
        return productRepository.findByCategoryHierarchy(categoryId);
    }

    public List<Product> getProductsByCategorySlugHierarchy(String slug) {
        return productRepository.findByCategorySlugHierarchy(slug);
    }
} 

File: services/product.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { Product } from '../models/product.interface';
import { environment } from '../../environments/environment';

@Injectable({
  providedIn: 'root'
})
export class ProductService {
  private apiUrl = `${environment.apiUrl}/products`;

  constructor(private http: HttpClient) { }

  // Tüm ürünleri getir
  getProducts(params?: {
    page?: number;
    size?: number;
    sort?: string;
    category?: string;
    search?: string;
  }): Observable<{ items: Product[], total: number }> {
    let httpParams = new HttpParams();
    
    if (params) {
      if (params.page) httpParams = httpParams.set('page', params.page.toString());
      if (params.size) httpParams = httpParams.set('size', params.size.toString());
      if (params.sort) httpParams = httpParams.set('sort', params.sort);
      if (params.category) httpParams = httpParams.set('category', params.category);
      if (params.search) httpParams = httpParams.set('search', params.search);
    }

    return this.http.get<{ items: Product[], total: number }>(this.apiUrl, { params: httpParams });
  }

  // Öne çıkan ürünleri getir
  getFeaturedProducts(): Observable<Product[]> {
    return this.http.get<Product[]>(`${this.apiUrl}/featured`);
  }

  // Tek bir ürün detayını getir
  getProduct(id: number): Observable<Product> {
    return this.http.get<Product>(`${this.apiUrl}/${id}`);
  }

  // Kategoriye göre ürünleri getir
  getProductsByCategory(categoryId: number | string): Observable<Product[]> {
    // Sayısal ID ile çağrı yap
    if (typeof categoryId === 'number' || !isNaN(Number(categoryId))) {
      const numericId = typeof categoryId === 'number' ? categoryId : Number(categoryId);
      return this.http.get<Product[]>(`${this.apiUrl}/category/${numericId}`);
    } 
    // String slug ile çağrı yap
    else {
      return this.http.get<Product[]>(`${this.apiUrl}/category/slug/${categoryId}`);
    }
  }

  // Ürün ara
  searchProducts(query: string): Observable<Product[]> {
    return this.http.get<Product[]>(`${this.apiUrl}/search`, {
      params: new HttpParams().set('q', query)
    });
  }

  // Yeni ürün ekle (Admin için)
  addProduct(product: Omit<Product, 'id'>): Observable<Product> {
    return this.http.post<Product>(this.apiUrl, product);
  }

  // Ürün güncelle (Admin için)
  updateProduct(id: number, product: Partial<Product>): Observable<Product> {
    return this.http.put<Product>(`${this.apiUrl}/${id}`, product);
  }

  // Ürün sil (Admin için)
  deleteProduct(id: number): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/${id}`);
  }

  // İndirimli ürünleri getir
  getDiscountedProducts(): Observable<Product[]> {
    return this.http.get<Product[]>(`${this.apiUrl}/discounted`);
  }

  // En çok satanları getir
  getBestSellers(): Observable<Product[]> {
    return this.http.get<Product[]>(`${this.apiUrl}/best-sellers`);
  }

  // Yeni ürünleri getir
  getNewArrivals(): Observable<Product[]> {
    return this.http.get<Product[]>(`${this.apiUrl}/new-arrivals`);
  }

  // Slug ile ürün getir
  getProductBySlug(slug: string): Observable<Product> {
    return this.http.get<Product>(`${this.apiUrl}/slug/${slug}`);
  }
}
